### PR TITLE
signature authentication for public links

### PR DESCRIPTION
**Pull request by C0rby**
Implemented the mechanism proposed here: https://github.com/cs3org/cs3apis/issues/110.
The signature authentication is limited to downloads.

### CODE REVIEW

**Comment by micbar** [comment]
@ishank011 we would like to bring that into the release. Could you do a fastlane 🏎️ review?

File: internal/grpc/services/publicshareprovider/publicshareprovider.go [hunk 1]
```diff
@@ -148,8 +149,17 @@ func (s *service) GetPublicShareByToken(ctx context.Context, req *link.GetPublic
 	log := appctx.GetLogger(ctx)
 	log.Debug().Msg("getting public share by token")
 
+	sig := req.GetAuthentication().GetSignature()
+	auth := publicshare.Authentication{
```
Comment (ishank011) [inline internal/grpc/services/publicshareprovider/publicshareprovider.go:153]: Can you just pass `req.GetAuthentication()` to the package? This looks like we're unnecessarily complicating things

File: internal/grpc/services/publicstorageprovider/publicstorageprovider.go [hunk 2]
```diff
@@ -552,6 +552,7 @@ func (s *service) ListContainer(ctx context.Context, req *provider.ListContainer
 	for i := range listContainerR.Infos {
 		filterPermissions(listContainerR.Infos[i].PermissionSet, ls.GetPermissions().Permissions)
 		listContainerR.Infos[i].Path = path.Join(s.mountPath, "/", tkn, relativePath, path.Base(listContainerR.Infos[i].Path))
+		addShare(listContainerR.Infos[i], ls)
```
Comment (ishank011) [inline internal/grpc/services/publicstorageprovider/publicstorageprovider.go]: Add error check

File: pkg/cbox/publicshare/sql/sql.go [hunk 3]
```diff
@@ -348,6 +353,9 @@ func (m *manager) ListPublicShares(ctx context.Context, u *user.User, filters []
 		if expired(cs3Share) {
 			_ = m.cleanupExpiredShares()
 		} else {
+			if cs3Share.PasswordProtected && sign {
```
Comment (ishank011) [inline pkg/cbox/publicshare/sql/sql.go:356]: This returns the password, not the sign

File: pkg/cbox/publicshare/sql/sql.go [hunk 4]
```diff
@@ -455,3 +468,18 @@ func checkPasswordHash(password, hash string) bool {
 	err := bcrypt.CompareHashAndPassword([]byte(strings.TrimPrefix(hash, "1|")), []byte(password))
 	return err == nil
 }
+
+func authenticate(share *link.PublicShare, pw string, auth publicshare.Authentication) bool {
+	switch {
+	case auth.Password != "":
+		return checkPasswordHash(auth.Password, pw)
+	case auth.Signature != publicshare.Signature{}:
```
Comment (ishank011) [inline pkg/cbox/publicshare/sql/sql.go:476]: Are you sure this condition isn't true when auth.Signature is nil?

**Review by ishank011 — CHANGES_REQUESTED** [review]
@C0rby you're returning the password instead of the sign in the getter methods.

File: pkg/cbox/publicshare/sql/sql.go (hunk 4 above)
Comment (C0rby) [reply pkg/cbox/publicshare/sql/sql.go:476]: `auth.Signature` can't be nil.

File: pkg/cbox/publicshare/sql/sql.go (hunk 3 above)
Comment (C0rby) [reply pkg/cbox/publicshare/sql/sql.go:356]: What do you mean? The `cs3Share.PasswordProtected`? It's a boolean which is true when a password is set for the share.

https://github.com/cs3org/reva/blob/538c8bdafb0ceeca96f650e580885dfeb4035276/pkg/cbox/utils/conversions.go#L213-L215
https://github.com/cs3org/reva/blob/538c8bdafb0ceeca96f650e580885dfeb4035276/pkg/cbox/utils/conversions.go#L236

Comment (ishank011) [reply pkg/cbox/publicshare/sql/sql.go:356]: Never mind

**Review by ishank011 — APPROVED** [review]

**Pull request merged**